src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,193 @@
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    {
+        size = 0;
+    }
+
+    private static class Node<T> {
+        private T item;
+        private Node<T> next;
+        private Node<T> prev;
+
+        Node(Node<T> prev, T element, Node<T> next) {
+            this.item = element;
+            this.next = next;
+            this.prev = prev;
+        }
+    }
+
     @Override
     public void add(T value) {
+        if (head == null) {
+            Node node = new Node<>(null, value, null);
+            head = node;
+            tail = head;
+        } else {
+            Node node = new Node<>(tail, value, null);
+            tail.next = node;
+            tail = node;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexIfAdd(index);
+        if (size == 0 && index == 0) {
+            Node node = new Node(null, value, null);
+            tail = node;
+            head = node;
+        } else if (size > 0 && index == 0) {
+            Node node = new Node(null, value, head);
+            head.prev = node;
+            head = node;
+        } else if (size > 0 && index == size) {
+            Node node = new Node(tail, value, null);
+            tail.next = node;
+            tail = node;
+        } else {
+            Node currentNode = indexIterator(index);
+            Node node = new Node(currentNode.prev, value, currentNode);
+            currentNode.prev.next = node;
+            currentNode.prev = node;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return (T) indexIterator(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node currentNode = indexIterator(index);
+        T tempValue = (T) currentNode.item;
+        currentNode.item = value;
+        return tempValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        if (size == 1) {
+            tail = null;
+            Node currentNode = head;
+            head = null;
+            size = 0;
+            return (T) currentNode.item;
+        }
+        if (size > 1 && index == 0) {
+            Node currentNode = head;
+            unlinkNodeFromHead(currentNode);
+            return (T) currentNode.item;
+        } else if (size > 1 && index == size - 1) {
+            Node currentNode = tail;
+            unlinkNodeFromTail(currentNode);
+            return (T) currentNode.item;
+        }
+        Node currentNode = indexIterator(index);
+        unlinkNode(currentNode);
+        return (T) currentNode.item;
```

Comment:
https://mate-academy.github.io/jv-program-common-mistakes/java-core/collections/linked-list.html#create-a-separate-method-unlinknode-node
Check common mistakes (Create a separate method)

Reply:
as you can see there are three separate methods for unlinking a node, depending where this node is situated